tendrl-monitoring-architecture.adoc

```diff
@@ -0,0 +1,197 @@
+    *** graphite-web - Graphite's user interface & API for rendering graphs
+        and dashboards
+  ** Metrics get fed into the stack via the Carbon service, which writes data
+     out to Whisper databases for long-term storage.
+  ** Advantages:
+    *** Graphite's web platform offers a variety of output styles and formats,
+        including raw images, CSV, XML, and JSON
+    *** Allows querying statistics for specific time-ranges and intervals.
+    *** Allows wild cards in query paths
+
+
+=== Collectd
+
+* A ​daemon​ which collects performance statistics periodically.
+* It is actively developed and supported and well documented.
+* Everything in collectd is done in plugins.
+  ** It comes with over 90 plugins​ and is extensible.
+  ** Collectd activates those plugins that are loaded and/or configured in
+     collectd’s conf files.
+  ** Changes in conf files will be effected in collectd only on restart of
+     the collectd daemon.
+* It serves the following responsibilities:
+  ** Collect the physical and logical resource utilizations using configured
+     plugins.
+  ** Watch for the breach of configured thresholds and trigger execution of
+     custom plugin.
+* Collectd comes packaged readily with a plugin to write collectd measured
+  statistics onto graphite db.
+
+
+Apart from the above, components of tendrl core stack also play important role
+in the functioning of tendrl monitoring stack.
+
+
+=== Tendrl Node Agent
+
+* The tendrl node agent opens a socket for collectd to send threshold breach
+  notifications to it.
+* The node agent then puts any threshold breach message from collectd onto
+  etcd.
+
+
+=== Tendrl SDS-Bridge
+
+* sds-bridge processes anything of interest to it on the node-agent exposed
+  threshold notification socket.
+* The sds-bridge also performs status monitoring.
+  ** In case of gluster, the gluster-bridge listens to and handles the events
+     that the gluster cluster provides.
+  ** In case of ceph, the ceph-bridge handles state changes in ceph resources
+
+
+== Generic Working Principle
+
+* As the final step of creating/managing an entity in tendrl, the appropriate
+  collectd plugin configurations will be made on suitable nodes in accordance
+  with the following:
+  ** At the start of tendrl performance monitoring application, default
+     monitoring configurations will be loaded to etcd.
+  ** The node-agent and the sds-bridge on any node started after this,
+     will use these default configurations for configuring collectd's plugins
+     (node-agent for physical resource plugins and bridge for bridge specific).
+    *** If the user chose intially not to have monitoring but later decided to
+        have it or node-agent/bridge was started before starting performance
+        monitoring application, a flow will be intiated internally to generate
+        collectd configurations once the performance monitoring application is
+        started.
```

Comment:
The monitoring application should be treated as an add-on stack. This means that the corresponding code, templates, configurations, flows etc. all need to be shipped for the monitoring stack should be shipped separately, with an assumption that the core stack is already started.

I would suggest that the flow framework from the core stack be reused for deploying the monitoring stack.
1. Install the monitoring application.
2. Monitoring application connects to etcd and uploads the default monitoring configuration and flows.
3. Monitoring application triggers the monitoring provisioning flow on every node for node monitoring via the core API.
4. The provisioning flow for the monitoring stack installs the collectd-related pieces on every node and generates the configuration files from templates shipped with the collectd plugins on nodes.